python/mxnet/pretrained_model.py

```diff
@@ -0,0 +1,674 @@
+
+    @property
+    def model(self):
+        return self._model
+
+    @property
+    def labels(self):
+        return self._labels
+
+    @property
+    def input_shape(self):
+        return self._input_shape
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def version(self):
+        return self._version
+
+    def __str__(self):
+        return "ImageClassifier: " + self._name + "(version %s)" % self._version
+
+    def __repr__(self):
+        return self.__str__()
+
+    def _make_dataiter(self, data, batch_size):
+        try:
+            import graphlab as _gl
+        except ImportError:
+            import sframe as _gl
+        except ImportError:
+            raise ImportError('Require GraphLab Create or SFrame')
+
+        if type(data) is not _gl.SFrame and type(data) is not _gl.SArray:
+            raise TypeError('Input data must be SFrame or SArray')
+        if type(data) is _gl.SArray and data.dtype() != _gl.Image:
+            raise TypeError('Expect image typed SArray, actual type is %s' % str(data.dtype()))
+        elif type(data) is _gl.SFrame:
+            if len(data.column_names()) != 1 or data.column_types()[0] != _gl.Image:
+                raise TypeError('Input SFrame must contain only a single image typed column')
+        if batch_size < len(data):
+            batch_size = len(data)
+
+        if type(data) is _gl.SArray:
+            data = _gl.SFrame({'image': data})
+        image_col = data.column_names()[0]
+        first_image = data[image_col][0]
+        input_shape = (first_image.height, first_image.width, first_image.channels)
+
+        if input_shape != tuple(self.input_shape):
+            _LOGGER.info('Detect image shape mismatches network input shape. Perform resize to shape %s' \
+                    % str(tuple(self.input_shape)))
+            data_resize = _gl.SFrame()
+            data_resize[image_col] = _gl.image_analysis.resize(data[image_col],
+                                                               self.input_shape[0],
+                                                               self.input_shape[1],
+                                                               self.input_shape[2], decode=True)
+            data = data_resize
+
+        dataiter = _io.SFrameImageIter(data, data_field=[image_col],
+                                       batch_size=batch_size,
+                                       mean_r=self.mean_rgb[0],
+                                       mean_g=self.mean_rgb[1],
+                                       mean_b=self.mean_rgb[2],
+                                       mean_nd=self.mean_nd,
+                                       data_name='data',
+                                       label_name='prob_label',
+                                       scale=self.rescale)
+        return dataiter
+
+
+    def extract_feature(self, data, batch_size=50):
+        # Make DataIter
```

Comment:
Also, maybe we can enable extracting features from multiple layers? Turns out global average pooling of last convolution layer gives more compact features which is quite good.